### PR TITLE
Add a new ConfigurationLoader interface

### PR DESCRIPTION
**Pull request by rmelick**
Add a new ConfigurationLoader interface to allow users to use their own fancier configuraiton loading that may support ecnrypted values or accessing properties from a tool like Hashicorp's Vault.  The current json loading from file now uses the new pattern.

### CODE REVIEW

**Comment by rmelick** [comment]
There has been no testing on this yet, just sending it out for feedback on the idea.

File: vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.util.Scanner;
+
+/**
+ *
+ */
+public class ConfigurationUtils {
+	private static final Logger log = LoggerFactory.getLogger(ConfigurationUtils.class);
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java]: If this is a constant (static final) variable and we want to be very strict regarding naming conventions, this should be called LOG

File: vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+				} catch (DecodeException e) {
+					log.error("Configuration file does not contain a valid JSON object");
+					return null;
+				}
+			} catch (FileNotFoundException e) {
+				log.error("Config file " + fileName + " does not exist");
+				return null;
+			}
+		} else {
+			conf = null;
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java]: I would throw an IllegalArgumentException here

File: vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+				String sconf = scanner.next();
+				try {
+					conf = new JsonObject(sconf);
+				} catch (DecodeException e) {
+					log.error("Configuration file does not contain a valid JSON object");
+					return null;
+				}
+			} catch (FileNotFoundException e) {
+				log.error("Config file " + fileName + " does not exist");
+				return null;
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java]: And I would throw this exception, so we do not swallow the exception

File: vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+		JsonObject conf;
+
+		if (fileName != null) {
+			try (Scanner scanner = new Scanner(new File(fileName)).useDelimiter("\\A")){
+				String sconf = scanner.next();
+				try {
+					conf = new JsonObject(sconf);
+				} catch (DecodeException e) {
+					log.error("Configuration file does not contain a valid JSON object");
+					return null;
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java]: And I would throw this exception, so we do not swallow the exception

File: vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+	public static JsonObject loadJsonFromFile(String fileName) {
+		JsonObject conf;
+
+		if (fileName != null) {
+			try (Scanner scanner = new Scanner(new File(fileName)).useDelimiter("\\A")){
+				String sconf = scanner.next();
+				try {
+					conf = new JsonObject(sconf);
+				} catch (DecodeException e) {
+					log.error("Configuration file does not contain a valid JSON object");
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java]: We should add the exception to the logger error so the stack trace is printed

File: vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+					return null;
+				}
+			} catch (FileNotFoundException e) {
+				log.error("Config file " + fileName + " does not exist");
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/configuration/ConfigurationUtils.java]: We should add the exception to the logger error so the stack trace is printed

File: vertx-platform/src/main/java/org/vertx/java/platform/impl/cli/Starter.java [hunk 7]
```diff
@@ -269,6 +271,15 @@ private void runBareHA(Args args) {
     block();
   }
 
+  private ConfigurationLoader getConfigurationLoader(String className) {
+    try {
+      return (ConfigurationLoader) Class.forName(className).newInstance();
+    } catch (InstantiationException | IllegalAccessException | ClassNotFoundException e) {
+      log.error("Could not instantiate configuration loader: " + className, e);
+      return null;
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/impl/cli/Starter.java]: And I would throw this exception, so we do not swallow the exception

File: vertx-platform/src/main/java/org/vertx/java/platform/impl/cli/Starter.java [hunk 8]
```diff
@@ -296,25 +307,16 @@ private void runVerticle(boolean zip, boolean module, String main, Args args) {
...
-      }
-    } else {
-      conf = null;
+    String configurationLoaderClass = args.map.get("-configurationLoader");
+    if (configurationLoaderClass == null) {
+      configurationLoaderClass = JsonFileConfigurationLoader.class.getName();
     }
+    ConfigurationLoader configurationLoader = getConfigurationLoader(configurationLoaderClass);
+    if (configurationLoader == null) {
+      return;
```
Comment (oscarfh) [inline vertx-platform/src/main/java/org/vertx/java/platform/impl/cli/Starter.java]: Should this ever be an acceptable outcome or it this an error that should be handled accordingly?

**Pull request closed without merging**